src/sentry/utils/arroyo_guard.py

```diff
@@ -0,0 +1,112 @@
+import logging
+from abc import ABC, abstractmethod
+from collections import deque
+from collections.abc import Callable
+from typing import Any, Deque, Generic, TypeVar
+
+from arroyo.dlq import InvalidMessage
+from arroyo.processing.strategies import MessageRejected
+from arroyo.processing.strategies.abstract import ProcessingStrategy
+from arroyo.types import Message, TStrategyPayload
+
+logger = logging.getLogger(__name__)
+
+TResult = TypeVar("TResult")
+TResult_co = TypeVar("TResult_co", covariant=True)
+
+
+class HasNextStep(ABC, Generic[TResult]):
+    @abstractmethod
+    def __init__(self, next_step: ProcessingStrategy[TResult], *args: Any, **kwargs: Any) -> None:
+        pass
+
+
+class NonFinalStrategy(ProcessingStrategy[TStrategyPayload], HasNextStep[TResult]):
+    pass
+
+
+def guard(
+    max_buffer_size: int = 10,
+) -> Callable[
+    [type[NonFinalStrategy[TStrategyPayload, TResult]]],
+    type[NonFinalStrategy[TStrategyPayload, TResult]],
+]:
+    """
+    Handles backpressure and invalid messages for any strategy. If
+    MessageRejected or InvalidMessage is received from the inner strategy
+    from the inner strategy the message is carried over by the guard for
+    future submission / reraise so the parent strategy does not have to implement
+    any handling for this scenario and can safely call `next_step.submit()`
+    in any part of the strategy.
```

Comment:
this i don't get. if the buffer reaches `max_buffer_size`, `submit` can still raise `MessageRejected`. you're not adding any guarantees with this buffer. the backpressure has to go somewhere: either it is propagated all the time or we OOM

Reply:
there's a couple of scenarios that come to mind:

- inner submit called outside the parent's submit method:. for example in https://github.com/getsentry/arroyo/blob/main/arroyo/processing/strategies/produce.py#L85 we raise MessageRejected on poll even though we are not supposed to
- a message passed to a parent strategy might be unbatched into more than one message to be forwarded to the next step downsteram. If one message in the middle fails with messagerejected/invalid message we need to resubmit only the second half of the batch not the whole thing. this is kinda the same thing we do in unfold today (https://github.com/getsentry/arroyo/blob/main/arroyo/processing/strategies/unfold.py#L37) where we put the remaining ones into a buffer, but this is an attempt to make that idea more reusable

Reply:
ah I misread the code. i think the intention makes sense.